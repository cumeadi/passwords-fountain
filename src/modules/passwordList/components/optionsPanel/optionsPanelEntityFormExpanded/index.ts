export { OptionsPanelEntityFormExpanded } from './optionsPanelEntityFormExpanded.component';
