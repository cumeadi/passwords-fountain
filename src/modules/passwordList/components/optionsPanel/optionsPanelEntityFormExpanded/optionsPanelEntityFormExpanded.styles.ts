import { h } from 'preact';
import { styled, setPragma } from 'goober';
import { theme } from '@/common/theme';

setPragma(h);

export const Wrapper = styled('section')``;

export const FormControlWrapper = styled('div')`
    padding-bottom: ${theme.spacing.m18};
`;

export const NoteLabelWrapper = styled('div')`
    color: ${theme.colors.darkBlue};
    font-style: italic;
    padding-top: ${theme.spacing.xs6};
`;
