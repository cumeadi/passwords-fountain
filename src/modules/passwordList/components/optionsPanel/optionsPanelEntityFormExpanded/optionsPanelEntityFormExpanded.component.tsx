import { h, VNode } from 'preact';
import { TypedComponent } from '@/common/typings/prop-types';
import PropTypes from 'prop-types';
import {
    Wrapper,
    FormControlWrapper,
} from './optionsPanelEntityFormExpanded.styles';
import { VariantProps } from '../optionsPanel.component';
import { variantNames } from '@/modules/passwordList/passwordList.contants';
import { FormControl } from '@/common/components/formControl';
import { ButtonWrapper, Content, ContentWrapper } from '../optionsPanel.styles';
import { Button } from '@/common/components/button';
import { Text } from '@/modules/localisation/components/text';
import { useRef, useState } from 'preact/hooks';
import { validateInputField } from '@/common/utils/form';
import { TextInput } from '@/common/components/textInput';
import { useAction } from '@/store';
import { passwordListActions } from '@/modules/passwordList/passwordList.actions';

const formValidation = {
    label(val?: string): boolean | string {
        return (val && val.length >= 4) || 'optionsPanel.labelTooShort';
    },
    login(val?: string): boolean | string {
        return (val && val.length >= 6) || 'optionsPanel.loginTooShort';
    },
    password(val?: string): boolean | string {
        return (val && val.length >= 4) || 'optionsPanel.passwordTooShort';
    },
    masterKey(val?: string) {
        return (val && val.length >= 6) || 'optionsPanel.masterKeyTooShort';
    },
} as const;

export const OptionsPanelEntityFormExpanded: TypedComponent<VariantProps> = ({
    switchCurrentVariantName,
}: VariantProps) => {
    const formRef = useRef(undefined as any);
    const addNewPassword = useAction(passwordListActions.addNewPassword);
    const [labelValue, setLabelValue] = useState('');
    const [labelErrors, setLabelErrors] = useState('');

    const [loginValue, setLoginValue] = useState('');
    const [loginErrors, setLoginErrors] = useState('');

    const [passwordValue, setPasswordValue] = useState('');
    const [passwordErrors, setPasswordErrors] = useState('');

    const [masterKeyValue, setMasterKeyValue] = useState('');
    const [masterKeyErrors, setMasterKeyErrors] = useState('');

    const handleCancelClick = (): void =>
        switchCurrentVariantName(variantNames.entityFormCollapsed);

    const handleAddClick = async (): Promise<void> => {
        await addNewPassword(
            {
                label: labelValue,
                login: loginValue,
                password: passwordValue,
            },
            masterKeyValue
        );
        switchCurrentVariantName(variantNames.entityFormCollapsed);
    };

    const renderLabelInput = (): VNode => (
        <TextInput
            placeholder="e.g. My Bank Account"
            hasError={Boolean(labelErrors)}
            name="label"
            value={labelValue}
            onInput={validateInputField(
                'label',
                formRef,
                formValidation,
                setLabelValue,
                setLabelErrors
            )}
        />
    );

    const renderLoginInput = (): VNode => (
        <TextInput
            placeholder="e.g. yourmail@yourmail.com"
            hasError={Boolean(loginErrors)}
            name="login"
            value={loginValue}
            onInput={validateInputField(
                'login',
                formRef,
                formValidation,
                setLoginValue,
                setLoginErrors
            )}
        />
    );

    const renderPasswordInput = (): VNode => (
        <TextInput
            placeholder="e.g. myPassWord1234"
            hasError={Boolean(passwordErrors)}
            name="password"
            value={passwordValue}
            onInput={validateInputField(
                'password',
                formRef,
                formValidation,
                setPasswordValue,
                setPasswordErrors
            )}
        />
    );

    const renderMasterKeyInput = (): VNode => (
        <TextInput
            placeholder="e.g. MyStrongPassword1234"
            hasError={Boolean(masterKeyErrors)}
            name="masterKey"
            value={masterKeyValue}
            onInput={validateInputField(
                'masterKey',
                formRef,
                formValidation,
                setMasterKeyValue,
                setMasterKeyErrors
            )}
        />
    );

    const isSubmitDisabled = [
        labelErrors,
        loginErrors,
        passwordErrors,
        masterKeyErrors,
    ].some(Boolean);
    const renderLabel = (label: string) => (): VNode => <Text>{label}</Text>;
    const renderError = (errors: string) => (): VNode => <Text>{errors}</Text>;
    return (
        <Wrapper>
            <ContentWrapper>
                <Content>
                    <form ref={formRef}>
                        <FormControlWrapper>
                            <FormControl
                                hasError={Boolean(labelErrors)}
                                renderLabel={renderLabel(
                                    'optionsPanel.labelInputLabel'
                                )}
                                renderInput={renderLabelInput}
                                renderError={renderError(labelErrors)}
                            />
                        </FormControlWrapper>
                        <FormControlWrapper>
                            <FormControl
                                hasError={Boolean(loginErrors)}
                                renderLabel={renderLabel(
                                    'optionsPanel.loginInputLabel'
                                )}
                                renderInput={renderLoginInput}
                                renderError={renderError(loginErrors)}
                            />
                        </FormControlWrapper>
                        <FormControlWrapper>
                            <FormControl
                                hasError={Boolean(passwordErrors)}
                                renderLabel={renderLabel(
                                    'optionsPanel.passwordInputLabel'
                                )}
                                renderInput={renderPasswordInput}
                                renderError={renderError(passwordErrors)}
                            />
                        </FormControlWrapper>
                        <FormControlWrapper>
                            <FormControl
                                hasError={Boolean(passwordErrors)}
                                renderLabel={renderLabel(
                                    'optionsPanel.enterMasterKey'
                                )}
                                renderInput={renderMasterKeyInput}
                                renderError={renderError(masterKeyErrors)}
                            />
                        </FormControlWrapper>
                    </form>
                </Content>
            </ContentWrapper>
            <ButtonWrapper>
                <Button onClick={handleCancelClick}>
                    <Text>optionsPanel.cancel</Text>
                </Button>
                <Button onClick={handleAddClick} disabled={isSubmitDisabled}>
                    <Text>optionsPanel.add</Text>
                </Button>
            </ButtonWrapper>
        </Wrapper>
    );
};

OptionsPanelEntityFormExpanded.propTypes = {
    switchCurrentVariantName: PropTypes.func.isRequired,
};
