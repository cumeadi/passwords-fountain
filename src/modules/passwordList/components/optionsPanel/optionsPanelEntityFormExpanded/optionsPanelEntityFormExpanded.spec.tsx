import '@testing-library/jest-dom/extend-expect';
import { h } from 'preact';
import { render } from '@testing-library/preact';
import { OptionsPanelEntityFormExpanded } from './optionsPanelEntityFormExpanded.component';
import { StoreProvider } from '@preact-hooks/unistore';
import createStore from 'unistore';

const stateMock = {
    passwordList: {
        selectedAndDecryptedEntity: {},
    },
};

describe('OptionsPanelAddNewExpanded', () => {
    it('should render correctly', () => {
        const { asFragment } = render(
            <StoreProvider value={createStore(stateMock)}>
                <OptionsPanelEntityFormExpanded
                    switchCurrentVariantName={(): void => {
                        /* eslint-disable @typescript-eslint/no-empty-function */
                    }}
                />
            </StoreProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
