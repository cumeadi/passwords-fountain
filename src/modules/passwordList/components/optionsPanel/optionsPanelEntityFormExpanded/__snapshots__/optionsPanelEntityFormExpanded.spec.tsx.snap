// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsPanelAddNewExpanded should render correctly 1`] = `
<DocumentFragment>
  <section
    class="go115679"
  >
    <div
      class="go115679"
    >
      <div
        class="go4038517794"
      >
        <form>
          <div
            class="go4038517794"
          >
            <section
              class="go3154290440"
            >
              <label
                class="go720639114"
                for="label"
              >
                Label
              </label>
              <div
                class="go3253370986"
              >
                <input
                  aria-describedby="label-error-message"
                  aria-invalid="false"
                  class="go3854179087"
                  id="label"
                  name="label"
                  placeholder="e.g. My Bank Account"
                  type="text"
                />
              </div>
            </section>
          </div>
          <div
            class="go4038517794"
          >
            <section
              class="go3154290440"
            >
              <label
                class="go720639114"
                for="login"
              >
                Login
              </label>
              <div
                class="go3253370986"
              >
                <input
                  aria-describedby="login-error-message"
                  aria-invalid="false"
                  class="go3854179087"
                  id="login"
                  name="login"
                  placeholder="e.g. yourmail@yourmail.com"
                  type="text"
                />
              </div>
            </section>
          </div>
          <div
            class="go4038517794"
          >
            <section
              class="go3154290440"
            >
              <label
                class="go720639114"
                for="password"
              >
                Password
              </label>
              <div
                class="go3253370986"
              >
                <input
                  aria-describedby="password-error-message"
                  aria-invalid="false"
                  class="go3854179087"
                  id="password"
                  name="password"
                  placeholder="e.g. myPassWord1234"
                  type="password"
                />
              </div>
            </section>
          </div>
          <div
            class="go4038517794"
          >
            <section
              class="go3154290440"
            >
              <label
                class="go720639114"
                for="encryptionKey"
              >
                Encryption key
                <div
                  class="go1162513423"
                >
                  Note: it's recommended to use master key here.
                </div>
              </label>
              <div
                class="go3253370986"
              >
                <input
                  aria-describedby="encryptionKey-error-message"
                  aria-invalid="false"
                  class="go3854179087"
                  id="encryptionKey"
                  name="encryptionKey"
                  placeholder="e.g. MyStrongPassword1234"
                  type="password"
                />
              </div>
            </section>
          </div>
          <input
            hidden=""
            type="submit"
          />
        </form>
      </div>
    </div>
    <div
      class="go3306721962"
    >
      <button
        class="go3135434547"
        type="button"
      >
        <span
          class="go4062099715"
        >
          Cancel
        </span>
      </button>
      <button
        class="go3135434547"
        disabled=""
        type="button"
      >
        <span
          class="go4062099715"
        >
          Add
        </span>
      </button>
    </div>
  </section>
</DocumentFragment>
`;
